src/bin/update-downloads.rs

```diff
@@ -31,401 +46,335 @@ fn main() {
     }
 }
 
-fn update(conn: &postgres::GenericConnection) -> postgres::Result<()> {
-    let mut max = 0;
-    loop {
-        // FIXME(rust-lang/rust#27401): weird declaration to make sure this
-        // variable gets dropped.
-        let tx;
-        tx = conn.transaction()?;
-        {
-            let stmt = tx.prepare(
-                "SELECT * FROM version_downloads \
-                                        WHERE processed = FALSE AND id > $1
-                                        ORDER BY id ASC
-                                        LIMIT $2",
-            )?;
-            let mut rows = stmt.query(&[&max, &LIMIT])?;
-            match collect(&tx, &mut rows)? {
-                None => break,
-                Some(m) => max = m,
-            }
-        }
-        tx.set_commit();
-        tx.finish()?;
+fn update(conn: &PgConnection) -> QueryResult<()> {
+    use version_downloads::dsl::*;
+    let mut max = Some(0);
+    while let Some(m) = max {
+        conn.transaction::<_, diesel::result::Error, _>(|| {
+            let rows = version_downloads
+                .filter(processed.eq(false))
+                .filter(id.gt(m))
+                .order(id)
+                .limit(LIMIT)
+                .load(conn)?;
+            collect(conn, &rows)?;
+            max = rows.last().map(|d| d.id);
+            Ok(())
+        })?;
     }
     Ok(())
 }
 
-fn collect(
-    tx: &postgres::transaction::Transaction,
-    rows: &mut postgres::rows::Rows,
-) -> postgres::Result<Option<i32>> {
+fn collect(conn: &PgConnection, rows: &[VersionDownload]) -> QueryResult<()> {
+    use diesel::{insert, update};
+
     // Anything older than 24 hours ago will be frozen and will not be queried
     // against again.
     let now = chrono::UTC::now();
     let cutoff = now.naive_utc().date() - chrono::Duration::days(1);
 
-    let mut map = HashMap::new();
-    for row in rows.iter() {
-        let download: VersionDownload = Model::from_row(&row);
-        assert!(map.insert(download.id, download).is_none());
-    }
```

Comment:
I am so happy this is getting cleaned up. There was some weird stuff going on in here.